pkg/apiclient/offline-client.go

```diff
@@ -4,76 +4,106 @@ import (
 	workflowarchivepkg "github.com/argoproj/argo-workflows/v3/pkg/apiclient/workflowarchive"
 	"github.com/argoproj/argo-workflows/v3/pkg/apiclient/workflowtemplate"
-	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 	"github.com/argoproj/argo-workflows/v3/workflow/templateresolution"
 
 	"sigs.k8s.io/yaml"
 )
 
+type offlineWorkflowTemplateGetterMap map[string]templateresolution.WorkflowTemplateNamespacedGetter
+
+func (m offlineWorkflowTemplateGetterMap) GetNamespaceGetter(namespace string) templateresolution.WorkflowTemplateNamespacedGetter {
+	v := m[namespace]
+	if v == nil {
+		return offlineWorkflowTemplateNamespacedGetter{
+			workflowTemplates: map[string]*wfv1.WorkflowTemplate{},
+			namespace:         namespace,
+		}
+	}
+
+	return m[namespace]
+}
+
 type offlineClient struct {
 	clusterWorkflowTemplateGetter       templateresolution.ClusterWorkflowTemplateGetter
-	namespacedWorkflowTemplateGetterMap map[string]templateresolution.WorkflowTemplateNamespacedGetter
+	namespacedWorkflowTemplateGetterMap offlineWorkflowTemplateGetterMap
 }
 
 var OfflineErr = fmt.Errorf("not supported when you are in offline mode")
 
 var _ Client = &offlineClient{}
 
-func newOfflineClient(files []string) (context.Context, Client, error) {
+// newOfflineClient creates a client that keeps all files (or files recursively contained within a path) given to it in memory.
+// It is useful for linting a set of files without having to connect to a cluster.
+func newOfflineClient(paths []string) (context.Context, Client, error) {
 	clusterWorkflowTemplateGetter := &offlineClusterWorkflowTemplateGetter{
 		clusterWorkflowTemplates: map[string]*wfv1.ClusterWorkflowTemplate{},
 	}
-	workflowTemplateGetters := map[string]templateresolution.WorkflowTemplateNamespacedGetter{}
+	workflowTemplateGetters := offlineWorkflowTemplateGetterMap{}
 
-	for _, file := range files {
-		bytes, err := os.ReadFile(file)
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to read file %s: %w", file, err)
-		}
-		var generic map[string]interface{}
-		if err := yaml.Unmarshal(bytes, &generic); err != nil {
-			return nil, nil, fmt.Errorf("failed to parse YAML from file %s: %w", file, err)
-		}
-		switch generic["kind"] {
-		case "ClusterWorkflowTemplate":
-			cwftmpl := new(v1alpha1.ClusterWorkflowTemplate)
-			if err := yaml.Unmarshal(bytes, &cwftmpl); err != nil {
-				return nil, nil, fmt.Errorf("failed to unmarshal file %s as a ClusterWorkflowTemplate: %w", file, err)
+	for _, basePath := range paths {
+		err := filepath.Walk(basePath, func(path string, info os.FileInfo, err error) error {
+			if info.IsDir() {
+				return nil
 			}
-
-			if _, ok := clusterWorkflowTemplateGetter.clusterWorkflowTemplates[cwftmpl.Name]; ok {
-				return nil, nil, fmt.Errorf("duplicate ClusterWorkflowTemplate found: %q", cwftmpl.Name)
+			if err != nil {
+				return err
 			}
```

Comment:
Reviewing without whitespace yields a nicer result here. The only real change is nestic the logic in a `filepath.Walk`